src/op_analytics/cli/subcommands/pulls/agora/delegates.py

```diff
@@ -0,0 +1,204 @@
+
+            print(f"max offset search range {lo} -- {hi}")
+
+            diff = hi - lo
+            midpoint = lo + (diff) // 2
+
+            if resp_lo.has_next and resp_hi.has_next:
+                lower_bound.add(hi)
+                lo = max(lower_bound)
+                hi = min(upper_bound) if upper_bound else hi * 10
+            elif resp_lo.has_next and not resp_hi.has_next:
+                lower_bound.add(lo)
+                upper_bound.add(hi)
+                if diff < 1000:
+                    self.min_offset = lo
+                    self.max_offset = hi
+                    return hi
+                else:
+                    hi = midpoint
+            else:
+                log.warning("Unexpected state encountered in binary search.")
+                hi = midpoint
+
+    def concurrent_fetch(self):
+        self.max_offset = self.binary_search_max_offset()
+        offsets = list(range(0, self.max_offset, self.limit))
+        data = run_concurrently(
+            function=lambda offset: self.request(offset).data,
+            targets=offsets,
+            max_workers=self.max_workers,
+        )
+        return list(itertools.chain.from_iterable(data.values()))
+
+
+def pull_delegates():
+    """Pull and write agora delegates data."""
+    p = Paginator(url=f"{BASE_URL}/delegates", params={"sort": "voting_power"}, max_workers=32)
+    delegate_data = p.concurrent_fetch()
+
+    # Clean and transform to Pandas
+    df = pd.json_normalize(delegate_data, sep="_")
+    df.columns = df.columns.map(_camelcase_to_snakecase)
+    df = df.convert_dtypes()
+    # In order to be able to drop duplicates we cannot have iterables in the dataframe
+    df = df.applymap(lambda x: str(x) if isinstance(x, list) else x)
+    # Check for duplicates
+    if df.shape[0] != len(df["address"].drop_duplicates()):
+        df = df.drop_duplicates()
+        log.warning(
+            f"Found duplicates in delegates data. Dropped {df.shape[0] - len(df['address'].drop_duplicates())} rows."
+        )
+
+    # Split to delegates that have voting power and those that don't
+    delegates_with_voting_power = df[df["voting_power_total"] > 0]
+    delegates_without_voting_power = df[df["voting_power_total"] == 0]
+
+    # Write to GCS
+    write(dataset=Agora.DELEGATES, dataframe=df, sort_by=["voting_power_total"])
+    # Why do this? Because the sizes are assymetrical and we want to optimize downstream
+    # performance.
+    write(
+        dataset=Agora.DELEGATES_WITH_VOTING_POWER,
+        dataframe=delegates_with_voting_power,
+        sort_by=["voting_power_total"],
+    )
+    write(
+        dataset=Agora.DELEGATES_WITHOUT_VOTING_POWER,
+        dataframe=delegates_without_voting_power,
+    )
+
+    return AgoraDelegates(
+        delegates_with_voting_power_df=delegates_with_voting_power,
+        delegates_without_voting_power_df=delegates_without_voting_power,
+    )
+
+
+# Todo: move to app.py
+def pull_delegate_data():
```

Comment:
I know we are pre-filtering for delegates with voting power > 0 for performance reasons, but just to confirm — we are still pulling data for delegates with voting power = 0, right?

I think it definitely makes sense to trim the delegate set down to those with non-zero voting power for votes, but think we'd still want to be able to track voting power balances over time for delegates whose voting power is 0 today (e.g. if a delegate had non-zero vp and voted on proposals last season, but then lost all their vp this season for whatever reason).

Reply:
Correct! We will be saving both cases but pulling events for delegates that have a non zero balance on a given date. 
An edge case to this, is if a delegate:
1. Gets voting power
2. Performs an event
3. Looses all the voting power before recorded in the delegates table

We would need to pull events for all delegates on a larger cadence to account for this, because Agora's API does not provide a way for us to pull recent votes and it should be time consuming to pull all the delegate information every day.

Reply:
Understood, thanks for clarifying! In that case, only extremely rare edge cases would be affected (i.e. intra-day swings from VP > 0 to VP = 0 coupled with a vote event on the same day), which we will eventually be able to fill in once we found a workaround solution for the limitations of the Agora API.